.terraform/test.sh

```diff
@@ -0,0 +1,129 @@
+#!/bin/bash
+#
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#####################
+# This script uses the Terraform configurations found in the subprojects
+# of this repository to provision a given GCP project with the correct
+# resources, perform integration testing, and then destroying the used
+# resources.
+#
+# Expected Environment Variables:
+#   TF_VAR_folder_id : Folder in which new GCP projects will be created
+#   TF_VAR_billing_account : Billing account to be used for created projects
+#   TF_VAR_project_prefix : Prefix to use for all created GCP projects
+#####################
+# Single Module Usage:
+# ./.terraform/test.sh <project> :: Tests only the given project
+#   ex: ./.terraform/test.sh java-accessapproval
+#
+# All Project Usage:
+# ./.terraform/test.sh :: Tests all projects in projects.txt
+
+function performTest() {
+  # Provision project with configuration found in './<proj>/.terraform' directory
+  pushd ./"$1"/.terraform || exit
+  terraform init || exit
+  terraform plan || exit
+  terraform apply -auto-approve || exit
+
+  # Use the project ID created by Terraform to set env and gcloud set-quota-project
+  GOOGLE_CLOUD_PROJECT=$(terraform output -raw project_id)
+  export GOOGLE_CLOUD_PROJECT
+  gcloud config set project "$GOOGLE_CLOUD_PROJECT"
+  gcloud auth application-default set-quota-project "$GOOGLE_CLOUD_PROJECT"
+
+  # Set module-specific environment variables for upcoming integration test(s)
+  if [[ -f "./env.sh" ]]; then
+    source "./env.sh"
+  fi
+
+  # Perform integration testing
+  pushd ../ || exit
+  mvn -B ${INTEGRATION_TEST_ARGS} \
+    -ntp \
+    -Penable-integration-tests \
+    -DtrimStackTrace=false \
+    -Dclirr.skip=true \
+    -Denforcer.skip=true \
+    -Dcheckstyle.skip=true \
+    -fae \
+    verify
+
+  exit_code=$?
+  popd || exit
```

Comment:
What's the situation where `popd` fails?

Reply:
I have this primarily due to `shellcheck` providing a warning otherwise. See https://github.com/koalaman/shellcheck/wiki/SC2164

> missing directories, missing permissions, broken symlinks